src/plots/cartesian/select.js

```diff
@@ -666,6 +666,15 @@ function isOnlyOnePointSelected(searchTraces) {
 function updateSelectedState(gd, searchTraces, eventData) {
     var i, searchInfo, cd, trace;
 
+    // before anything else, update preGUI if necessary
+    for(i = 0; i < searchTraces.length; i++) {
+        var fullInputTrace = searchTraces[i].cd[0].trace._fullInput;
+        var tracePreGUI = gd._fullLayout._tracePreGUI[fullInputTrace.uid];
+        if(tracePreGUI.selectedpoints === undefined) {
+            tracePreGUI.selectedpoints = fullInputTrace._input.selectedpoints || null;
+        }
+    }
```

Comment:
Missed this one earlier as it doesn't go through `restyle` or emit a `plotly_restyle` event - but anyway it wants to be a special case so that we don't push individual array elements into `_tracePreGUI`, we link the array as a unit.

BTW I'm working off `trace._fullInput` here but I notice farther down we go straight from `trace` to `trace._input` - have we ever looked at selections on `groupby` traces?

Reply:
Actually it looks like selections work fine with groupby 😌 there's some potential confusion around `pointNumber` vs `pointIndex` in the event data, but all the information is there (`curveNumber` and `pointIndex` give you the index in the input arrays, `pointNumber` is the index within the group), and visible behavior is all correct.

The only thing that looks odd to me is `fullTrace.selectedpoints`:
- Immediately after drawing it has the selected input array indices from items in that group, but `fullTrace._fullInput.selectedpoints` is missing
- After you make an edit (`restyle` or zoom, for example) `fullTrace.selectedpoints` and `fullTrace._fullInput.selectedpoints` *both* have the entire array of input array indices.

I don't see any practical effects of this difference, though it's possible `Plotly.react`'s diffing algo will care.

Anyway I think I've found a `uirevision` + `selectedpoints` bug, investigating.

Reply:
> Anyway I think I've found a `uirevision` + `selectedpoints` bug, investigating.

haha looks like it's just the missing `fullInput.selectedpoints` I mentioned above - fix coming, and perhaps I'll make `fullTrace.selectedpoints` self-consistent across edits as well while I'm at it.

Reply:
alright, 🔒 in https://github.com/plotly/plotly.js/pull/3236/commits/090231bce09a0ee064b0741ec261b9b1d10de9b3 - which should also speed things up a bit in the normal non-transform case where `data` and `fullData` can point to the same array (as they will after an edit anyway) so we only need to build it once.